fairscale/nn/data_parallel/sharded_ddp.py

```diff
@@ -116,122 +185,180 @@ def no_sync(self) -> Generator:
-            if not self.accumulate_grads:
-                self.need_reduction = True
-            if self.broadcast_buffers and len(list(self.module.buffers())) > 0:
-                self._sync_buffers()
-
-        return self.module(*inputs, **kwargs)
-
-    def reduce(self) -> None:
-        """
-        This function must be called explicitly after backward to reduce
-        gradients. There is no automatic hook like c10d.
-        """
-        assert self.module.training, "Cannot call reduce in eval"
-
-        if not self.need_reduction or self.accumulate_grads:
-            return
-
-        self.need_reduction = False
-
-        with torch.no_grad():
-            for device, per_device in self.sharded_optimizer.per_device_params.items():
-                self._reduce_grads_task(
-                    self._reduce_buffers[device],
-                    per_device,
-                    group=self.process_group,
-                    self_rank=self.rank,
-                    world_size=self.world_size,
-                )
-
     @staticmethod
-    def _reduce_grads_task(
-        buffers: List[torch.Tensor], per_rank_params: List[List[Parameter]], group: Any, self_rank: int, world_size: int
-    ) -> None:
-        """Helper to reduce a list of params. The params are sorted by size, smallest first, which allows for
-        an opportunistic bucketing.
-
-        NOTE: All param gradients are assumed to exist"""
-
-        buffer_size = buffers[0].numel()
-        bucket_requests = []
-        requests = []
-
-        for (rank, params), buffer in zip(enumerate(per_rank_params), buffers):
-            # All the params are sorted per rank and per increasing size
-            if len(params) == 0:
-                continue
-
-            for p in params:
-                if p.grad is None:
-                    p.grad = torch.zeros_like(p)
-
-            global_rank = OSS.get_global_rank(group, rank)
-
-            # Copy small gradients into per-GPU buffers and then async reduce
-            i_bucketed = 0  # the number of tensors packed in the buffer
-            offset = 0
-
-            # Since all the parameters are already sorted per increasing size, we only need to consider the first ones.
-            while i_bucketed < len(params) and offset + params[i_bucketed].numel() < buffer_size:
-                end = offset + params[i_bucketed].numel()
-                buffer[offset:end].copy_(params[i_bucketed].grad.data.view(-1))  # type: ignore
-                offset = end
-                i_bucketed += 1
-
-            if i_bucketed > 0:
-                buffer.div_(world_size)
-                bucket_requests.append(
-                    (
-                        dist.reduce(tensor=buffer, dst=global_rank, group=group, async_op=True),  # type: ignore
-                        rank,
-                    )
-                )
+    def _mark_inputs_requires_grad(*inputs: Any) -> Any:
```

Comment:
This will not work for non-float tensors.

Reply:
add a docstring on why this function is needed?

Reply:
@msbaines @min-xu-ai essentially this needs 1 tensor to be marked, so that the Gatekeeper is registered in the autograd graph. I would really love to get rid of that, other solutions mentioned [here for instance](https://discuss.pytorch.org/t/custom-autograd-function-backward-pass-not-called/9144) seem deprecated